File: core/orchestrator/ticker_relevance.py
"""Ticker Relevance — deterministic relevance scorer.

Given a crisis domain, entities, country, and keywords, selects the most
relevant subset of the stock universe (max 30 tickers + up to 3 indices).

This is a pure, deterministic scorer — no LLM calls. It replaces the old
approach of hardcoding ticker lists per scenario.

Selection pipeline:
1. **Entity resolution** — resolve named orgs to tickers via UniverseLoader
2. **Domain→sector mapping** — map crisis domain to relevant sectors
3. **Country/region filter** — prioritise tickers from the affected country/region
4. **Keyword tag match** — boost tickers whose tags overlap with keywords
5. **Beta regime** — determine which beta regime (IT, US, EM, default) applies
6. **Cap at 30** — rank by relevance score, take top 30
"""

from __future__ import annotations

import logging
from dataclasses import dataclass, field

from core.orchestrator.financial_impact import UniverseLoader

logger = logging.getLogger(__name__)

# ── Domain → Sector Mapping ──────────────────────────────────────────────────

DOMAIN_SECTORS: dict[str, list[str]] = {
    "political": ["banking", "sovereign_debt", "insurance", "utilities", "defense"],
    "financial": ["banking", "insurance", "sovereign_debt", "real_estate", "energy_fossil", "utilities", "tech", "automotive"],
    "commercial": ["tech", "food_consumer", "luxury", "automotive", "media"],
    "corporate": ["tech", "automotive", "energy_fossil", "infrastructure", "luxury"],
    "geopolitical": ["defense", "energy_fossil", "banking", "sovereign_debt"],
    "social": ["media", "telecom", "healthcare", "food_consumer"],
    "environmental": ["energy_fossil", "energy_renewable", "utilities", "automotive"],
    "health": ["healthcare", "food_consumer", "insurance"],
    "technology": ["tech", "telecom", "media"],
}

# ── Country → Beta Regime ────────────────────────────────────────────────────

COUNTRY_TO_REGIME: dict[str, str] = {
    "IT": "IT", "US": "US",
    # Emerging markets
    "BR": "EM", "IN": "EM", "CN": "EM", "MX": "EM", "AR": "EM",
    "TW": "EM", "KR": "EM", "CL": "EM", "HK": "EM",
    # Developed Europe → default (no special regime)
    "DE": "default", "FR": "default", "GB": "default",
    "NL": "default", "CH": "default", "DK": "default",
    "SE": "default", "FI": "default", "JP": "default", "AU": "default",
}

# Broad country text → regime mapping for free-form input
_COUNTRY_TEXT_TO_REGIME: dict[str, str] = {
    "italy": "IT", "italia": "IT", "italian": "IT",
    "united states": "US", "usa": "US", "america": "US", "american": "US",
    "brazil": "EM", "brasil": "EM", "india": "EM", "china": "EM",
    "mexico": "EM", "argentina": "EM", "taiwan": "EM", "korea": "EM",
    "chile": "EM", "hong kong": "EM",
    "germany": "default", "france": "default", "uk": "default",
    "united kingdom": "default", "japan": "default", "australia": "default",
    "netherlands": "default", "switzerland": "default", "denmark": "default",
    "sweden": "default", "finland": "default",
}

# ── Country → Region ─────────────────────────────────────────────────────────

COUNTRY_TO_REGION: dict[str, str] = {
    "IT": "europe", "DE": "europe", "FR": "europe", "GB": "europe",
    "NL": "europe", "CH": "europe", "DK": "europe", "SE": "europe",
    "FI": "europe",
    "US": "north_america",
    "JP": "asia_pacific", "KR": "asia_pacific", "CN": "asia_pacific",
    "HK": "asia_pacific", "TW": "asia_pacific", "IN": "asia_pacific",
    "AU": "asia_pacific",
    "BR": "latam", "MX": "latam", "AR": "latam", "CL": "latam",
}


# ── Data Classes ─────────────────────────────────────────────────────────────

@dataclass
class RelevantUniverse:
    """The selected subset of tickers + indices for a scenario."""
    tickers: list[dict] = field(default_factory=list)   # [{ticker, name, sector, score, ...}]
    indices: list[dict] = field(default_factory=list)    # [{ticker, name, country}]
    beta_regime: str = "default"
    rationale: str = ""


# ── Scorer ───────────────────────────────────────────────────────────────────

class TickerRelevanceScorer:
    """Deterministic ticker relevance scorer.

    Usage:
        scorer = TickerRelevanceScorer()
        result = scorer.select("financial", ["UniCredit"], "IT", ["banking"])
    """

    def __init__(self):
        self._loader = UniverseLoader()

    def _resolve_regime(self, country: str) -> str:
        """Map a country string to a beta regime."""
        if not country:
            return "default"
        # Try direct ISO code
        upper = country.strip().upper()
        if upper in COUNTRY_TO_REGIME:
            return COUNTRY_TO_REGIME[upper]
        # Try free-form text
        lower = country.strip().lower()
        if lower in _COUNTRY_TEXT_TO_REGIME:
            return _COUNTRY_TEXT_TO_REGIME[lower]
        return "default"

    def _resolve_country_code(self, country: str) -> str | None:
        """Map free-form country to ISO code for region filtering."""
        if not country:
            return None
        upper = country.strip().upper()
        if upper in COUNTRY_TO_REGIME:
            return upper
        lower = country.strip().lower()
        # Reverse lookup
        text_to_code = {
            "italy": "IT", "italia": "IT", "italian": "IT",
            "united states": "US", "usa": "US", "america": "US", "american": "US",
            "brazil": "BR", "brasil": "BR", "india": "IN", "china": "CN",
            "mexico": "MX", "argentina": "AR", "taiwan": "TW", "korea": "KR",
            "chile": "CL", "hong kong": "HK",
            "germany": "DE", "france": "FR", "uk": "GB",
            "united kingdom": "GB", "japan": "JP", "australia": "AU",
            "netherlands": "NL", "switzerland": "CH", "denmark": "DK",
            "sweden": "SE", "finland": "FI",
        }
        return text_to_code.get(lower)

    def select(
        self,
        domain: str,
        entities: list[str],
        country: str,
        keywords: list[str],
        *,
        max_tickers: int = 30,
        max_indices: int = 3,
    ) -> RelevantUniverse:
        """Select the most relevant tickers for a scenario.

        Args:
            domain: Crisis domain (political, financial, commercial, etc.)
            entities: Named organisations (e.g. ["Apple", "Samsung"])
            country: Country code or name (e.g. "IT", "Brazil")
            keywords: Topic keywords (e.g. ["smartphone", "ai"])
            max_tickers: Cap on returned tickers (default 30)
            max_indices: Cap on returned indices (default 3)

        Returns:
            RelevantUniverse with scored tickers, indices, beta regime, rationale.
        """
        regime = self._resolve_regime(country)
        country_code = self._resolve_country_code(country)
        region = COUNTRY_TO_REGION.get(country_code, "") if country_code else ""

        # 1. Entity resolution → direct matches (highest priority)
        direct_tickers: dict[str, float] = {}  # ticker → score
        for entity in entities:
            resolved = self._loader.resolve_org(entity.lower())
            for t in resolved:
                direct_tickers[t] = direct_tickers.get(t, 0) + 10.0

        # 2. Domain → sector expansion
        domain_lower = domain.lower().strip()
        relevant_sectors = DOMAIN_SECTORS.get(domain_lower, [])

        # 3. Score all stocks
        scored: dict[str, float] = dict(direct_tickers)
        all_stocks = self._loader._data.get("stocks", [])
        keyword_set = {k.lower() for k in keywords}

        for stock in all_stocks:
            ticker = stock["ticker"]
            score = scored.get(ticker, 0.0)

            # Sector relevance
            if stock.get("sector") in relevant_sectors:
                sector_idx = relevant_sectors.index(stock["sector"])
                score += 5.0 - (sector_idx * 0.3)  # primary sectors score higher

            # Country match
            if country_code and stock.get("country") == country_code:
                score += 3.0
            elif region and stock.get("region") == region:
                score += 1.0

            # Keyword tag match
            tags = set(stock.get("tags", []))
            overlap = tags & keyword_set
            score += len(overlap) * 2.0

            # Market cap boost (mega > large > mid)
            cap_tier = stock.get("market_cap_tier", "mid")
            if cap_tier == "mega":
                score += 1.0
            elif cap_tier == "large":
                score += 0.5

            if score > 0:
                scored[ticker] = score

        # 4. Ensure minimum sector diversity (at least 3 sectors)
        if relevant_sectors and len(relevant_sectors) < 3:
            # Add fallback sectors
            fallback = ["banking", "tech", "energy_fossil", "healthcare", "utilities"]
            for fb in fallback:
                if fb not in relevant_sectors:
                    relevant_sectors.append(fb)
                    if len(relevant_sectors) >= 5:
                        break
            # Re-score with fallback sectors (lower weight)
            for stock in all_stocks:
                ticker = stock["ticker"]
                if ticker not in scored and stock.get("sector") in relevant_sectors:
                    scored[ticker] = 1.0

        # Ensure at least 3 sectors represented in output.
        # Strategy: first pick best-in-sector for each relevant sector (diversity),
        # then fill remaining slots by score.
        sorted_tickers = sorted(scored.items(), key=lambda x: -x[1])
        min_sectors = 3

        selected: list[dict] = []
        selected_set: set[str] = set()
        sectors_seen: set[str] = set()

        def _make_entry(ticker: str, score: float) -> dict | None:
            stock = self._loader.get_stock(ticker)
            if stock is None:
                return None
            return {
                "ticker": ticker,
                "name": stock.get("name", ""),
                "sector": stock.get("sector", ""),
                "country": stock.get("country", ""),
                "score": round(score, 2),
            }

        # Phase 1: ensure diversity — pick top ticker per sector
        sector_best: dict[str, tuple[str, float]] = {}
        for ticker, score in sorted_tickers:
            stock = self._loader.get_stock(ticker)
            if stock is None:
                continue
            sector = stock.get("sector", "")
            if sector and sector not in sector_best:
                sector_best[sector] = (ticker, score)

        # Add best from each relevant sector first
        for sector in relevant_sectors:
            if sector in sector_best and len(selected) < max_tickers:
                ticker, score = sector_best[sector]
                if ticker not in selected_set:
                    entry = _make_entry(ticker, score)
                    if entry:
                        selected.append(entry)
                        selected_set.add(ticker)
                        sectors_seen.add(sector)

        # Phase 2: fill remaining slots by score
        for ticker, score in sorted_tickers:
            if len(selected) >= max_tickers:
                break
            if ticker in selected_set:
                continue
            entry = _make_entry(ticker, score)
            if entry:
                selected.append(entry)
                selected_set.add(ticker)
                sectors_seen.add(entry["sector"])

        # Phase 3: if still fewer than 3 sectors, add top stocks from other sectors
        if len(sectors_seen) < min_sectors:
            all_sector_names = set(s.get("sector") for s in all_stocks)
            missing = all_sector_names - sectors_seen
            for sector in sorted(missing):
                if len(sectors_seen) >= min_sectors:
                    break
                sector_stocks = self._loader.tickers_for_sector(sector)
                if sector_stocks:
                    top = sector_stocks[0]
                    if top["ticker"] not in selected_set:
                        selected.append({
                            "ticker": top["ticker"],
                            "name": top.get("name", ""),
                            "sector": sector,
                            "country": top.get("country", ""),
                            "score": 0.5,
                        })
                        selected_set.add(top["ticker"])
                        sectors_seen.add(sector)

        # Trim to max
        selected = selected[:max_tickers]

        # 5. Select indices
        indices = self._select_indices(country_code, region, max_indices)

        # 6. Build rationale
        entity_str = ", ".join(entities) if entities else "none"
        keyword_str = ", ".join(keywords) if keywords else "none"
        sector_str = ", ".join(sorted(sectors_seen))
        rationale = (
            f"Domain: {domain}. Entities: {entity_str}. "
            f"Country: {country or 'unspecified'} (regime: {regime}). "
            f"Keywords: {keyword_str}. "
            f"Sectors covered: {sector_str}. "
            f"Selected {len(selected)} tickers and {len(indices)} indices."
        )

        return RelevantUniverse(
            tickers=selected,
            indices=indices,
            beta_regime=regime,
            rationale=rationale,
        )

    def _select_indices(
        self, country_code: str | None, region: str, max_indices: int
    ) -> list[dict]:
        """Pick the most relevant market indices."""
        all_indices = self._loader._data.get("indices", [])
        selected = []

        # Priority: country-specific index first
        for idx in all_indices:
            if country_code and idx.get("country") == country_code:
                selected.append(idx)

        # Then regional
        for idx in all_indices:
            if idx not in selected and region and idx.get("region") == region:
                selected.append(idx)

        # Then global benchmarks (S&P 500 always useful)
        global_benchmarks = {"^GSPC", "^STOXX50E"}
        for idx in all_indices:
            if idx not in selected and idx.get("ticker") in global_benchmarks:
                selected.append(idx)

        return selected[:max_indices]

File: frontend/lib/ticker-relevance.ts
/**
 * Ticker Relevance System
 *
 * Selects relevant stock tickers for a simulation scenario based on
 * domain, geography, mentioned entities, and keyword tags.
 */

/* ------------------------------------------------------------------ */
/*  Types                                                              */
/* ------------------------------------------------------------------ */

export interface BetaCoefficients {
  political_beta: number;
  spread_beta: number;
  crisis_alpha: number;
  volatility_multiplier: number;
}

export interface SectorDef {
  label: string;
  gics_code: string;
  betas: Record<string, BetaCoefficients>;
}

export interface StockEntry {
  ticker: string;
  name: string;
  country: string;
  region: string;
  sector: string;
  industry_group: string;
  market_cap_tier: string;
  index_membership: string[];
  tags: string[];
}

export interface IndexEntry {
  ticker: string;
  name: string;
  country: string;
  region: string;
}

export interface StockUniverse {
  version: string;
  updated: string;
  indices: IndexEntry[];
  stocks: StockEntry[];
  sectors: Record<string, SectorDef>;
  org_aliases: Record<string, string[]>;
}

export interface SelectedTicker {
  ticker: string;
  name: string;
  sector: string;
  reason: string;
}

export interface TickerSelection {
  tickers: SelectedTicker[];
  indices: IndexEntry[];
  betaRegime: string;
}

/* ------------------------------------------------------------------ */
/*  Geography -> Beta Regime                                           */
/* ------------------------------------------------------------------ */

const COUNTRY_TO_REGIME: Record<string, string> = {
  Italy: "IT",
  IT: "IT",
  US: "US",
  "United States": "US",
  Brazil: "EM",
  BR: "EM",
  India: "EM",
  IN: "EM",
  China: "EM",
  CN: "EM",
  Mexico: "EM",
  MX: "EM",
  Argentina: "EM",
  AR: "EM",
  Chile: "EM",
  CL: "EM",
  "South Korea": "EM",
  KR: "EM",
  Taiwan: "EM",
  TW: "EM",
  Germany: "US",
  DE: "US",
  France: "US",
  FR: "US",
  UK: "US",
  GB: "US",
  Japan: "US",
  JP: "US",
};

/** Map a geography string (country name or ISO code) to a beta regime key. */
function resolveRegime(geography: string): string {
  if (!geography) return "default";
  return COUNTRY_TO_REGIME[geography] ?? "default";
}

/* ------------------------------------------------------------------ */
/*  Domain -> preferred sectors                                        */
/* ------------------------------------------------------------------ */

const DOMAIN_SECTORS: Record<string, string[]> = {
  political: [
    "banking",
    "defense",
    "energy_fossil",
    "sovereign_debt",
    "media",
  ],
  financial: ["banking", "insurance", "real_estate", "sovereign_debt", "energy_fossil", "utilities", "tech", "automotive"],
  commercial: ["tech", "automotive", "food_consumer", "luxury", "telecom"],
  corporate: ["tech", "automotive", "healthcare", "infrastructure", "defense"],
  energy: [
    "energy_fossil",
    "energy_renewable",
    "utilities",
    "infrastructure",
  ],
  health: ["healthcare", "food_consumer", "insurance"],
  military: ["defense", "infrastructure", "tech"],
};

/* ------------------------------------------------------------------ */
/*  Public API                                                         */
/* ------------------------------------------------------------------ */

/**
 * Look up beta coefficients for a sector + regime.
 * Falls back to "default" regime if the requested one is missing.
 */
export function getBeta(
  universe: StockUniverse,
  sector: string,
  regime: string,
): BetaCoefficients {
  const sectorDef = universe.sectors[sector];
  if (!sectorDef) {
    return {
      political_beta: 1.0,
      spread_beta: 0,
      crisis_alpha: 0,
      volatility_multiplier: 1.0,
    };
  }
  return sectorDef.betas[regime] ?? sectorDef.betas["default"];
}

/**
 * Resolve an organisation name to ticker(s) via the alias map.
 * Case-insensitive lookup.
 */
export function resolveOrg(universe: StockUniverse, org: string): string[] {
  const key = org.toLowerCase();
  return universe.org_aliases[key] ?? [];
}

/**
 * Select the most relevant tickers for a scenario.
 *
 * Strategy:
 *  1. Direct entity matches (highest priority)
 *  2. Tag-matched stocks (keywords matched against stock tags)
 *  3. Domain-sector fill (sector preferences per domain)
 *  4. Diversity guarantee (min 3 sectors)
 *  5. Cap at 30 tickers
 *  6. Pick 1-3 indices by geography
 */
export function selectRelevantTickers(
  universe: StockUniverse,
  domain: string,
  entities: string[],
  geography: string,
  keywords: string[],
): TickerSelection {
  const betaRegime = resolveRegime(geography);
  const selected = new Map<string, SelectedTicker>();

  // 1. Direct entity matches
  for (const entity of entities) {
    const tickers = resolveOrg(universe, entity);
    for (const t of tickers) {
      const stock = universe.stocks.find((s) => s.ticker === t);
      if (stock && !selected.has(t)) {
        selected.set(t, {
          ticker: t,
          name: stock.name,
          sector: stock.sector,
          reason: `entity:${entity}`,
        });
      }
    }
  }

  // 2. Tag-matched stocks
  if (keywords.length > 0) {
    const lowerKw = keywords.map((k) => k.toLowerCase());
    for (const stock of universe.stocks) {
      if (selected.size >= 30) break;
      if (selected.has(stock.ticker)) continue;
      const stockTags = stock.tags.map((t) => t.toLowerCase());
      const match = lowerKw.some((kw) => stockTags.includes(kw));
      if (match) {
        selected.set(stock.ticker, {
          ticker: stock.ticker,
          name: stock.name,
          sector: stock.sector,
          reason: `tag:${keywords.find((k) => stockTags.includes(k.toLowerCase()))}`,
        });
      }
    }
  }

  // 3. Domain-sector fill
  const preferredSectors = DOMAIN_SECTORS[domain] ?? Object.keys(universe.sectors).slice(0, 5);
  for (const sector of preferredSectors) {
    if (selected.size >= 30) break;
    const sectorStocks = universe.stocks
      .filter((s) => s.sector === sector && !selected.has(s.ticker))
      .sort((a, b) => {
        const tierOrder: Record<string, number> = {
          mega: 0,
          large: 1,
          mid: 2,
          small: 3,
        };
        return (tierOrder[a.market_cap_tier] ?? 4) - (tierOrder[b.market_cap_tier] ?? 4);
      });
    // Take top 2-3 per sector
    const take = Math.min(3, sectorStocks.length, 30 - selected.size);
    for (let i = 0; i < take; i++) {
      const stock = sectorStocks[i];
      selected.set(stock.ticker, {
        ticker: stock.ticker,
        name: stock.name,
        sector: stock.sector,
        reason: `domain:${domain}/${sector}`,
      });
    }
  }

  // 4. Diversity guarantee: ensure at least 3 distinct sectors
  const currentSectors = new Set(
    Array.from(selected.values()).map((t) => t.sector),
  );
  if (currentSectors.size < 3) {
    const allSectors = Object.keys(universe.sectors);
    for (const sector of allSectors) {
      if (currentSectors.size >= 3) break;
      if (currentSectors.has(sector)) continue;
      if (selected.size >= 30) break;
      const stock = universe.stocks.find(
        (s) => s.sector === sector && !selected.has(s.ticker),
      );
      if (stock) {
        selected.set(stock.ticker, {
          ticker: stock.ticker,
          name: stock.name,
          sector: stock.sector,
          reason: `diversity:${sector}`,
        });
        currentSectors.add(sector);
      }
    }
  }

  // 5. Cap at 30
  const tickers = Array.from(selected.values()).slice(0, 30);

  // 6. Select 1-3 indices by geography
  const indices = selectIndices(universe, geography, betaRegime);

  return { tickers, indices, betaRegime };
}

/* ------------------------------------------------------------------ */
/*  Internal helpers                                                   */
/* ------------------------------------------------------------------ */

function selectIndices(
  universe: StockUniverse,
  geography: string,
  regime: string,
): IndexEntry[] {
  const result: IndexEntry[] = [];

  // Try to find a local index
  const countryCode = geography.length === 2 ? geography : undefined;
  if (countryCode) {
    const local = universe.indices.find((i) => i.country === countryCode);
    if (local) result.push(local);
  }

  // Map regime to primary index
  if (regime === "IT") {
    const mib = universe.indices.find((i) => i.country === "IT");
    if (mib && !result.some((r) => r.ticker === mib.ticker)) result.push(mib);
    const stoxx = universe.indices.find((i) => i.country === "EU");
    if (stoxx && !result.some((r) => r.ticker === stoxx.ticker))
      result.push(stoxx);
  } else if (regime === "US") {
    const sp = universe.indices.find((i) => i.ticker === "^GSPC");
    if (sp && !result.some((r) => r.ticker === sp.ticker)) result.push(sp);
  } else if (regime === "EM") {
    // Find matching EM country index
    const countryMap: Record<string, string> = {
      Brazil: "BR",
      BR: "BR",
      India: "IN",
      IN: "IN",
      China: "CN",
      CN: "CN",
      Mexico: "MX",
      MX: "MX",
      "South Korea": "KR",
      KR: "KR",
    };
    const cc = countryMap[geography];
    if (cc) {
      const idx = universe.indices.find((i) => i.country === cc);
      if (idx && !result.some((r) => r.ticker === idx.ticker))
        result.push(idx);
    }
  }

  // Always include S&P 500 as global reference if not yet present
  if (!result.some((r) => r.ticker === "^GSPC")) {
    const sp = universe.indices.find((i) => i.ticker === "^GSPC");
    if (sp) result.push(sp);
  }

  return result.slice(0, 3);
}
